server/session/session.go

```diff
@@ -0,0 +1,164 @@
+package session
+
+import (
+	pluginapi "github.com/mattermost/mattermost-plugin-api"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-apps/apps"
+	"github.com/mattermost/mattermost-plugin-apps/server/proxy/request"
+	"github.com/mattermost/mattermost-plugin-apps/server/store"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+)
+
+const (
+	sessionLengthInMinutes    = 10
+	minSessionLengthInMinutes = 5
+)
+
+type Service interface {
+	GetOrCreate(appID apps.AppID, userID string) (*model.Session, error)
+	ListForUser(userID string) ([]*model.Session, error)
+	RevokeSessionsForApp(c *request.Context, appID apps.AppID) error
+}
+
+var _ Service = (*service)(nil)
+
+type service struct {
+	log   utils.Logger
+	mm    *pluginapi.Client
+	store *store.Service
+}
+
+func NewService(mm *pluginapi.Client, store *store.Service) Service {
+	return &service{
+		log:   utils.NewPluginLogger(mm),
+		mm:    mm,
+		store: store,
+	}
+}
+
+func (s *service) GetOrCreate(appID apps.AppID, userID string) (*model.Session, error) {
+	session, err := s.store.Session.Get(appID, userID)
+
+	if err == nil && !session.IsExpired() {
+		err = s.extendSessionExpiryIfNeeded(appID, userID, session)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to extend session length")
+		}
+
+		return session, nil
+	}
+
+	if err != nil && !errors.Is(err, utils.ErrNotFound) {
+		return nil, errors.Wrap(err, "failed to get session from store")
+	}
+
+	return s.createSession(appID, userID)
+}
+
+func (s *service) createSession(appID apps.AppID, userID string) (*model.Session, error) {
+	user, err := s.mm.User.Get(userID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch user for new session")
+	}
+
+	app, err := s.store.App.Get(appID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch app for new session")
+	}
+
+	session := &model.Session{
+		UserId:    userID,
+		Roles:     user.Roles,
+		IsOAuth:   true,
+		ExpiresAt: model.GetMillis() + (1000 * 60 * sessionLengthInMinutes),
+	}
+	session.GenerateCSRF()
+
+	// TODO: The buit-in app also needs some (?) of this props. Does it also need an OAuth app?
```

Comment:
I would argue that the built-in app uses the plugin API and should not require nor create oauthApps or sessions.

Reply:
Interesting idea. I like the idea of using the RPC API for expansions. @levb Thoughts?

Reply:
Expansions are really tricky. Using sessions seems fine to have the "permissions" for free, but should we use the App sessions then, and not the "plugin session"? That would mean that for using expansions, we would need an OauthApp and a session in any case, with the scopes to read that particular piece.

Reply:
In order to create an OauthApp an actingUser is required. Given that such OauthApp needs to be created when the plugin starts, I don't think we can follow that road.

I ended up going the "plugin session" route by using the RPC client for the builtin app when doing expansions.